Add `ClassValueCompat` to support systems without `java.lang.ClassValue` (such as Android)

nwk37011 opened the pull request:
The goal of this commit is to enable systems without java.lang.ClassValue to fall back to other method which does not utilize java.lang.ClassValue while keeping its current behavior as how it is for systems with java.lang.ClassValue. Use of java.lang.ClassValue seems mainly for caching as https://github.com/scala/scala/pull/7879 introduced, so I ran a related benchmark from one of the commits and got following results

```
Baseline

[info] Benchmark                                             Mode  Cnt   Score    Error   Units
[info] ClassTagBenchmark.lookupClassTag                      avgt   20  40.328 ±  0.496   ns/op
[info] ClassTagBenchmark.lookupClassTag:·gc.alloc.rate       avgt   20  ≈ 10⁻⁴           MB/sec
[info] ClassTagBenchmark.lookupClassTag:·gc.alloc.rate.norm  avgt   20  ≈ 10⁻⁵             B/op
[info] ClassTagBenchmark.lookupClassTag:·gc.count            avgt   20     ≈ 0           counts
```

```
ClassValueCompat - java.lang.ClassValue exists

[info] Benchmark                                             Mode  Cnt   Score    Error   Units
[info] ClassTagBenchmark.lookupClassTag                      avgt   20  42.834 ±  0.231   ns/op
[info] ClassTagBenchmark.lookupClassTag:·gc.alloc.rate       avgt   20  ≈ 10⁻⁴           MB/sec
[info] ClassTagBenchmark.lookupClassTag:·gc.alloc.rate.norm  avgt   20  ≈ 10⁻⁵             B/op
[info] ClassTagBenchmark.lookupClassTag:·gc.count            avgt   20     ≈ 0           counts
```

For cases where systems lacking java.lang.ClassValue, I earned the following result which seems obviously slower than the above two but still it enables such systems to actually utilize classes such as ClassTag instead of throwing ClassNotFoundException

```
ClassValueCompat - java.lang.ClassValue does not exist

[info] Benchmark                                                      Mode  Cnt     Score    Error   Units
[info] ClassTagBenchmark.lookupClassTag                               avgt   20    79.452 ±  0.294   ns/op
[info] ClassTagBenchmark.lookupClassTag:·gc.alloc.rate                avgt   20  3838.471 ± 15.265  MB/sec
[info] ClassTagBenchmark.lookupClassTag:·gc.alloc.rate.norm           avgt   20   336.000 ±  0.001    B/op
[info] ClassTagBenchmark.lookupClassTag:·gc.churn.G1_Eden_Space       avgt   20  3843.122 ± 24.785  MB/sec
[info] ClassTagBenchmark.lookupClassTag:·gc.churn.G1_Eden_Space.norm  avgt   20   336.408 ±  1.792    B/op
[info] ClassTagBenchmark.lookupClassTag:·gc.churn.G1_Old_Gen          avgt   20     0.003 ±  0.001  MB/sec
[info] ClassTagBenchmark.lookupClassTag:·gc.churn.G1_Old_Gen.norm     avgt   20    ≈ 10⁻⁴             B/op
[info] ClassTagBenchmark.lookupClassTag:·gc.count                     avgt   20  1525.000           counts
[info] ClassTagBenchmark.lookupClassTag:·gc.time                      avgt   20   655.000               ms
```

This commit combined with https://github.com/scala/scala/pull/9739 will open Scala 2.13.x to wider range of systems including Android as discussed and demonstrated at https://users.scala-lang.org/t/scala-2-13-on-android/7235/24

smarter commented:
Nice! FWIW, I've also opened an issue against Android about the lack of ClassValue, but it's not clear to me if they're interested in fixing it: https://issuetracker.google.com/issues/196063118 (maybe someone with some twitter clout could try bringing attention to this?)

lrytz commented on src/library/scala/runtime/ClassValueCompat.java:
```diff
@@ -0,0 +1,66 @@
+
+public abstract class ClassValueCompat<T> {
+
+    private static final boolean classValueAvailable = checkClassValueAvailability();
```
does it make sense to keep this field? it's only read once in te constructor.

nwk37011 replied:
No, agree it as no reason to be kept

lrytz commented on src/library/scala/runtime/ClassValueCompat.java:
```diff
@@ -0,0 +1,66 @@
+    private final ClassValueInterface<T> instance;
+
+    protected ClassValueCompat() {
+        instance = classValueAvailable ? new JavaClassValue() : new FallbackClassValue();
```
```suggestion
        instance = checkClassValueAvailability() ? new JavaClassValue() : new FallbackClassValue();
```

lrytz commented on src/library/scala/runtime/ClassValueCompat.java:
```diff
@@ -0,0 +1,66 @@
+
+    private static boolean checkClassValueAvailability() {
+        try {
+            Class.forName("java.lang.ClassValue", false, ClassValueCompat.class.getClassLoader());
```
maybe use Object's class loader, as the two should be defined in the same library?

```suggestion
            Class.forName("java.lang.ClassValue", false, Object.class.getClassLoader());
```

lrytz commented on src/library/scala/runtime/ClassValueCompat.java:
```diff
@@ -0,0 +1,66 @@
+
+package scala.runtime;
+
+public abstract class ClassValueCompat<T> {
```
We cannot add a public class in a Scala 2.13.x release due to forwards binary compatibility.

I think the only workaround here is to write the class in Scala (not Java) and mark it `private[scala]`. This requires a [MiMa filter](https://github.com/scala/scala/blob/2.13.x/project/MimaFilters.scala); running `library/mimaReportBinaryIssues` in sbt will tell you what to put in there.

I see that `ClassValueCompat` is used in the Java class `ModuleSerializationProxy` and we're using `CompileOrder.JavaThenScala`, so that's a problem. I guess we can create a private copy of `ClassValueCompat` within `ModuleSerializationProxy`.

smarter replied:
> We cannot add a public class in a Scala 2.13.x release due to forwards binary compatibility.

Does this restriction applies to classes in scala.runtime too? /cc @sjrd

lrytz replied:
I think so. `scala.reflect.internal` is excluded, but that's only in the "reflect" module, we need the class in the "library" module.

sjrd replied:
> 
> 
> > We cannot add a public class in a Scala 2.13.x release due to forwards binary compatibility.
> 
> Does this restriction applies to classes in scala.runtime too? /cc @sjrd

The thing is that this class shouldn't be in `scala.runtime`. `scala.runtime` is supposed to be for code called by the codegen. Not for code called from library code. Code called from library code should be in the library, next to what calls them. There is no reason to put this class in `scala.runtime`.

lrytz replied:
I'm not sure we need to be so strict about this, there's other stuff in `scala.runtime` too (RichInt).

sjrd replied:
> 
> 
> I'm not sure we need to be so strict about this, there's other stuff in `scala.runtime` too (RichInt).

Yes, and it's not great that they are there, as it muddies the waters between what's public API and what isn't. I had a WiP PR to move them out, but I didn't finish it in time for 2.13.0.

nwk37011 replied:
Relocated `ClassValueCompat` to be under `scala.compat`, rewritten in Scala and applied the suggestions above mentioned in the comments, thanks for the directions

SethTisue replied:
`scala.compat` currently only has a single deprecated class in it. So that package was well on its way to being eliminated entirely, but this PR would bring it back to life.

I might be more comfortable just putting the new class in `scala.runtime`, since doing it that way is the status quo. Then in the future if someone wants to submit a PR like @sjrd's that reorganizes and/or splits `scala.runtime` according to some guiding principle, then great.

@sjrd you objected to using `scala.runtime`, but you didn't (?) propose a better location. wdyt of the choice of `scala.compat`?

sjrd replied:
My suggestion is to put it in whatever actual package it is used, as package-private to that package.

nwk37011 replied:
It's currently being utilized for `scala.runtime` and `scala.reflect`, will there be a way to declare as package-private of two packages?

NthPortal replied:
`private[scala]` (the most specific super-package of `scala.runtime` and `scala.reflect`), and probably in the `scala` package. there's some argument to having a `scala.internal` package for this and the sorts of things @sjrd mentioned that are internal but not generated by the compiler, but I'm not sure if it's worth investing in creating that package outside of a major release or not.

nwk37011 replied:
`ClassValueCompat` Relocated as `private[scala]` under `scala` package as suggested

lrytz reviewed (commented):
Thanks for that contribution, the approach LGTM!

bishabosha commented on src/library/scala/compat/ClassValueCompat.scala:
```diff
@@ -0,0 +1,47 @@
+    true
+  }.getOrElse(false)
+
+  private val instance = if (checkClassValueAvailability()) new JavaClassValue() else new FallbackClassValue()
```
just a question as I am unfamiliar with patching standard library like this: is the JVM and Android lazy enough not to error reading `java.lang.ClassValue` in the parents of `JavaClassValue` before executing `checkClassValueAvailability`?

dwijnand replied:
Yeah, fairly sure the class isn't loaded until it's first seen.

dwijnand commented on src/library/scala/runtime/ModuleSerializationProxy.java:
```diff
@@ -48,4 +46,56 @@ public ModuleSerializationProxy(Class<?> moduleClass) {
+        }
+    }
+
+    private interface ClassValueInterface<T> {
```
Do we have to have a Java version and a Scala version?  And I mean of all of it (including checkClassValueAvailability, ClassValueCompat and its children).

nwk37011 replied:
Made an attempt to convert ModuleSerializationProxy.java to Scala but I am not very sure if this will cause any kind of ill side-effects for not knowing what exactly it does.

sideeffffect commented on src/library/scala/runtime/ModuleSerializationProxy.scala:
```diff
@@ -0,0 +1,47 @@
+  }
+
+  private def rethrowRuntime(e: Throwable): Object = {
+    val cause = e.getCause
```
Twice `.getCause` seems a bit suspicious (one in the definition of `rethrowRuntime` and one when it is used). If it's intentional, it would benefit from a comment explaining the reason. Thank you! :smiley:

lrytz commented on src/library/scala/ClassValueCompat.scala:
```diff
@@ -0,0 +1,47 @@
+
+import scala.util.Try
+
+private[scala] abstract class ClassValueCompat[T] {
```
Sorry to bring this to discussion again, but do we really want to add this class to the `scala` package? Even if it's package private it will show up in some places way too prominently. For example when editing a Java file in a mixed project, `scala.ClassValueCompat` will be a public class.

I'd rather keep it in `scala.runtime`.

SethTisue replied:
agree; I argued (at least somewhat) against `scala.compat`, but I think `scala` is an even worse choice

nwk37011 replied:
Issues and fixes mentioned below and here got applied to the latest commit

lrytz commented on src/library/scala/ClassValueCompat.scala:
```diff
@@ -0,0 +1,47 @@
+private[scala] abstract class ClassValueCompat[T] {
+
+  private def checkClassValueAvailability(): Boolean = Try {
+    Class.forName("java.lang.ClassValue", false, getClass.getClassLoader)
```
```suggestion
    Class.forName("java.lang.ClassValue", false, classOf[Object].getClassLoader)
```

lrytz commented on src/library/scala/ClassValueCompat.scala:
```diff
@@ -0,0 +1,47 @@
+
+private[scala] abstract class ClassValueCompat[T] {
+
+  private def checkClassValueAvailability(): Boolean = Try {
```
use `try` and only `catch` the expected exception?

lrytz commented on src/library/scala/ClassValueCompat.scala:
```diff
@@ -0,0 +1,47 @@
+
+  private trait ClassValueInterface {
+    def get(param1Class: Class[_]): T
+    def remove(param1Class: Class[_]): Unit
```
not that it matters much, but the name `param1Class` is a bit strange. Maybe just use `cls`, and also use `cls` instead of `` `type` `` in the other methdos?

lrytz commented on src/library/scala/runtime/ModuleSerializationProxy.scala:
```diff
@@ -0,0 +1,47 @@
+  private val instances = new ClassValueCompat[Object] {
+    override protected def computeValue(`type`: Class[_]): Object = {
+      try {
+        java.security.AccessController.doPrivileged(new PrivilegedExceptionAction[Object] {
+          override def run(): Object = {
+            `type`.getField("MODULE$").get(null)
+          }
+        })
```
```suggestion
        java.security.AccessController.doPrivileged((() => `type`.getField("MODULE$").get(null)): PrivilegedExceptionAction[Object])
```

lrytz commented on src/library/scala/runtime/ModuleSerializationProxy.scala:
```diff
@@ -0,0 +1,47 @@
+
+@SerialVersionUID(1L)
+final class ModuleSerializationProxy(moduleClass: Class[_]) extends Serializable {
+  @SuppressWarnings(Array("unused")) private def readResolve = ModuleSerializationProxy.instances.get(moduleClass)
```
`SuppressWarnings` doesn't do anything in Scala code

lrytz reviewed (commented):
I think the rewrite of `ModuleSerializationProxy` in Scala looks good, but it's certainly not trivial to think of all the possible consequences.

sideeffffect commented on src/library/scala/runtime/ClassValueCompat.scala:
```diff
@@ -0,0 +1,49 @@
+  }
+
+  private class FallbackClassValue extends ClassValueInterface {
+    override def get(cls: Class[_]): T = ClassValueCompat.this.get(cls)
```
If `java.lang.ClassValue` isn't present, doesn't `ClassValueCompat#get` result in an endless loop between `ClassValueCompat#get` and `FallbackClassValue#get`?

nwk37011 replied:
oops, yes it has to be `computeValue`, got it fixed, thank you

sideeffffect commented on src/library/scala/runtime/ClassValueCompat.scala:
```diff
@@ -0,0 +1,49 @@
+
+package scala.runtime
+
+private[scala] abstract class ClassValueCompat[T] {
+
+  private def checkClassValueAvailability(): Boolean =
+    try {
+      Class.forName("java.lang.ClassValue", false, classOf[Object].getClassLoader)
+      true
+    } catch {
+      case _: ClassNotFoundException => false
+    }
+
+  private val instance = if (checkClassValueAvailability()) new JavaClassValue() else new FallbackClassValue()
+
+  private class JavaClassValue extends ClassValue[T] with ClassValueInterface {
+    override def computeValue(cls: Class[_]): T = ClassValueCompat.this.computeValue(cls)
+  }
+
+  private class FallbackClassValue extends ClassValueInterface {
+    override def get(cls: Class[_]): T = ClassValueCompat.this.computeValue(cls)
+
+    override def remove(cls: Class[_]): Unit = {}
+  }
+
+  private trait ClassValueInterface {
+    def get(cls: Class[_]): T
+
+    def remove(cls: Class[_]): Unit
+  }
+
+  def get(cls: Class[_]): T = instance.get(cls)
+
+  def remove(cls: Class[_]): Unit = instance.remove(cls)
+
+  protected def computeValue(cls: Class[_]): T
+
+}
```
Do you think it would be better to have `instance` in constructor?
```suggestion

import scala.runtime.ClassValueCompat._

private[scala] abstract class ClassValueCompat[T]() extends ClassValueInterface[T] { self =>

  private val instance =
    try { new JavaClassValue() }
    catch { case _: ClassNotFoundException => new FallbackClassValue() }

  private class JavaClassValue extends ClassValue[T] with ClassValueInterface[T] {
    override def computeValue(cls: Class[_]): T = self.computeValue(cls)
  }

  private class FallbackClassValue extends ClassValueInterface[T] {
    override def get(cls: Class[_]): T = self.computeValue(cls)

    override def remove(cls: Class[_]): Unit = {}
  }

  def get(cls: Class[_]): T = instance.get(cls)

  def remove(cls: Class[_]): Unit = instance.remove(cls)

  protected def computeValue(cls: Class[_]): T

}

private[scala] object ClassValueCompat {

  trait ClassValueInterface[T] {
    def get(cls: Class[_]): T
    def remove(cls: Class[_]): Unit
  }

}
```

nwk37011 replied:
I am not exactly sure about the need of constructor, in which cases do we benefit from having it?

sideeffffect replied:
1. Having `instance` in constructor seemed conceptually simpler to me
 2. `catch { case _: ClassNotFoundException => new FallbackClassValue() }` seemed like an elegant way to construct the object, while handling the possibility of `ClassValue` missing
 3. I like when all `ClassValueCompat`, `JavaClassValue` and `FallbackClassValue` implement the same interface `ClassValueInterface`

But all these things are just stylistic details (maybe except for `3.`, it would be good design to expose a single interface to the rest of the system) that are internal so go with what you feel is best.

nwk37011 replied:
I cannot make this code to compile, compiler tells me `JavaClassValue` and `FallbackClassValue` missing, what am I missing here?

sideeffffect replied:
You're missing imports.
```
import com.github.ccpcrypto.ClassValueCompat._
```
above `private[scala] abstract class ClassValueCompat...`.
Or change to `ClassValueCompat.JavaClassValue` and `ClassValueCompat.FallbackClassValue`.

Generally speaking, IntelliJ IDEA should help you with imports (it can suggest what to import or do the imports automatically, etc). Anyway, I've amended my suggestion to contain the required import.

sideeffffect replied:
@nwk37011 I have to apologize for mystifying you :bow: Indeed you're right that the code I've presented doesn't compile because of "missing" `JavaClassValue` and `FallbackClassValue`, it seems directly related to the alternative constructor I used.
I've changed the code to use private field `instance` and it compiles now.

lrytz committed:
ClassValueCompat to support systems without java.lang.ClassValue

On runtime systems where `java.lang.ClassValue` is not available
`ClassValueCompat` re-computes the value on each `get` invocation.

Co-authored-by: nwk37011 <nwk37011@gmail.com>

lrytz reviewed (approved):
Rebased and squashed.

Thank you, @nwk37011!

sideeffffect commented on src/library/scala/runtime/ClassValueCompat.scala:
```diff
@@ -0,0 +1,53 @@
+    def remove(cls: Class[_]): Unit
+  }
+
+  private val classValueAvailable: Boolean = try {
```
Would it in any way be beneficial to make it `lazy val`? :thinking:

lrytz replied:
I don't think so, the field is only initialized when the enclosing object is first accessed.

SethTisue commented:
on JDK 17 we're seeing (at https://app.travis-ci.com/github/scala/scala/jobs/542786102):

```
[error] /home/travis/build/scala/scala/src/library/scala/runtime/ModuleSerializationProxy.scala:23:23: class AccessController in package security is deprecated
[error]         java.security.AccessController.doPrivileged((() => cls.getField("MODULE$").get(null)): PrivilegedExceptionAction[Object])
[error]                       ^
```

PR with fix: #9785

fiatjaf commented:
I have a question related to this PR, so I thought I would just beg for attention here. Let me know if this is inappropriate: https://stackoverflow.com/questions/73042201/why-a-scala-2-13-app-is-trying-to-use-javaclassvalue-on-android-if-it-is-not-a
